File: tools/wave/src/database/sessions-database.js
const path = require("path");
const DataStore = require("nedb");

const DatabaseUtils = require("../utils/database-utils");
const Deserializer = require("../utils/deserializer");
const Serializer = require("../utils/serializer");
const JobQueue = require("../utils/job-queue");
const Session = require("../data/session");

const { promisifyNedbDataStore } = DatabaseUtils;
const DEFAULT_FILE_PATH = "./sessions.db";
const DEFAULT_COMPACTION_INTERVAL = 60000;

const READ_JOB_GROUP = "read";
const MAX_ACCESS_JOBS = 1;
const MAX_GROUP_JOBS = 5;

class SessionsDatabase {
  constructor({ compactionInterval = DEFAULT_COMPACTION_INTERVAL } = {}) {
    this._compactionInterval = compactionInterval;
    this._sessionsAccessQueue = new JobQueue(MAX_ACCESS_JOBS, {
      groupLimit: MAX_GROUP_JOBS
    });
    this._queueSessionsAccess = this._sessionsAccessQueue.queueJob.bind(
      this._sessionsAccessQueue
    );
  }

  async initialize({
    filePath = DEFAULT_FILE_PATH,
    resultsDatabase,
    testsDatabase
  } = {}) {
    let sessionsDataStore = new DataStore({
      filename: filePath
    });

    sessionsDataStore.persistence.setAutocompactionInterval(
      this._compactionInterval
    );
    sessionsDataStore = promisifyNedbDataStore(sessionsDataStore);
    await sessionsDataStore.loadDatabase();
    this._db = sessionsDataStore;
    this._resultsDatabase = resultsDatabase;
    this._testsDatabase = testsDatabase;
  }

  async createSession(session) {
    return this._queueSessionsAccess(() => this._createSession(session));
  }

  async _createSession(session) {
    const token = session.getToken();
    const sessionObject = Serializer.serializeSession(session);

    await this._resultsDatabase.loadDatabase(token);

    const { COMPLETED, ABORTED } = Session;
    if (session.getStatus() !== COMPLETED && session.getStatus() !== ABORTED) {
      await this._testsDatabase.loadDatabase(token);
      const { pending_tests, running_tests, completed_tests } = sessionObject;
      await this._testsDatabase.createTests(token, {
        pending_tests,
        running_tests,
        completed_tests
      });
    }

    delete sessionObject.completed_tests;
    delete sessionObject.running_tests;
    delete sessionObject.pending_tests;
    await this._db.insert(sessionObject);
  }

  async readSession(token) {
    return this._queueSessionsAccess(() => this._readSession(token), {
      group: READ_JOB_GROUP
    });
  }

  async _readSession(token) {
    const result = await this._db.find({ token });
    if (!result || result.length === 0) {
      return null;
    }
    const session = Deserializer.deserializeSession(result[0]);
    if(session.sessionLabel){
    console.log("**DEBUG** _ SESSION LABEL:", session.sessionLabel);
  }
    const { COMPLETED, ABORTED } = Session;
    if (session.getStatus() !== COMPLETED && session.getStatus() !== ABORTED) {
      await this._testsDatabase.loadDatabase(token);
      const tests = await this._testsDatabase.readTests(token);
      if (tests) {
        const { pending_tests, running_tests, completed_tests } = tests;
        if (pending_tests) session.setPendingTests(pending_tests);
        if (completed_tests) session.setCompletedTests(completed_tests);
        if (running_tests) session.setRunningTests(running_tests);
      }
    }

    return session;
  }

  async readSessions() {
    return this._queueSessionsAccess(() => this._readSessions(), {
      group: READ_JOB_GROUP
    });
  }

  async _readSessions() {
    const result = await this._db.find({});
    if (!result) {
      return [];
    }
    return Deserializer.deserializeSessions(result);
  }

  async readPublicSessions() {
    return this._queueSessionsAccess(() => this._readPublicSessions(), {
      group: READ_JOB_GROUP
    });
  }

  async _readPublicSessions() {
    const result = await this._db.find({ is_public: true });
    if (!result) {
      return [];
    }
    return Deserializer.deserializeSessions(result);
  }

  async updateSession(session) {
    return this._queueSessionsAccess(() => this._updateSession(session));
  }

  async _updateSession(session) {
    const token = session.getToken();
    const result = await this._db.find({ token });
    if (!result || result.length === 0) {
      return null;
    }
    const sessionObject = Serializer.serializeSession(session);
    await this._resultsDatabase.loadDatabase(token);

    const { COMPLETED, ABORTED } = Session;
    if (session.getStatus() !== COMPLETED && session.getStatus() !== ABORTED) {
      await this._testsDatabase.loadDatabase(token);
      const { pending_tests, running_tests, completed_tests, sessionLabel } = sessionObject;
      await this._testsDatabase.updateTests(token, {
        pending_tests,
        running_tests,
        completed_tests
      });
    } else {
      await this._testsDatabase.deleteTests(token);
    }

    delete sessionObject.completed_tests;
    delete sessionObject.running_tests;
    delete sessionObject.pending_tests;
    await this._db.update({ token }, sessionObject);
  }

  async deleteSession(token) {
    return this._queueSessionsAccess(() => this._deleteSession(token));
  }

  async _deleteSession(token) {
    const sessionDataStore = this._db;
    await sessionDataStore.remove({ token }, true);
    await this._testsDatabase.deleteTests(token);
    await this._resultsDatabase.deleteResults(token);
  }

  async findTokens(fragment) {
    return this._queueSessionsAccess(() => this._findTokens(fragment), {
      group: READ_JOB_GROUP
    });
  }

  async _findTokens(fragment) {
    const results = await this._db.find({
      token: new RegExp("^" + fragment)
    });
    return results.map(result => result.token);
  }
}

module.exports = SessionsDatabase;

File: tools/wptserve/wptserve/wave/database/sessions_database.py
from __future__ import absolute_import
import re

from ..data.session import COMPLETED, ABORTED
from ..utils.serializer import serialize_session
from ..utils.deserializer import deserialize_session, deserialize_sessions

class SessionsDatabase(object):
    def initialize(self, results_database, tests_database):
        self._results_database = results_database
        self._tests_database = tests_database

        #tmp
        self._sessions = []

    def create_session(self, session):
        token = session.token
        session_dict = serialize_session(session)

        tests = {}
        if session.status != COMPLETED and session.status != ABORTED:
            tests[u"pending_tests"] = session.pending_tests
            tests[u"running_tests"] = session.running_tests
            tests[u"completed_tests"] = session.completed_tests

        tests[u"malfunctioning_tests"] = session.malfunctioning_tests
        self._tests_database.create_tests(token, tests)

        del session_dict[u"completed_tests"]
        del session_dict[u"running_tests"]
        del session_dict[u"pending_tests"]
        del session_dict[u"malfunctioning_tests"]
        self._sessions.append(session_dict)

    def read_session(self, token):
        session_dict = None
        for session in self._sessions:
            if session[u"token"] == token:
                session_dict = session
                break

        if session_dict is None: return None

        session = deserialize_session(session_dict)

        tests = self._tests_database.read_tests(token)

        if tests is not None:
            if session.status != COMPLETED and session.status != ABORTED:
                if u"pending_tests" in tests: session.pending_tests = tests[u"pending_tests"]
                if u"completed_tests" in tests: session.completed_tests = tests[u"completed_tests"]
                if u"running_tests" in tests: session.running_tests = tests[u"running_tests"]
            if u"malfunctioning_tests" in tests: session.malfunctioning_tests = tests[u"malfunctioning_tests"]

        return session

    def read_sessions(self):
        return deserialize_sessions(self._sessions)

    def read_expiring_sessions(self):
        expiring_sessions = []
        for session in self._sessions:
            if session[u"expiration_date"] == None: continue
            expiring_sessions.append(session)
        expiring_sessions = deserialize_sessions(expiring_sessions)
        return expiring_sessions

    def read_public_sessions(self):
        public_sessions = []
        for session in self._sessions:
            if not session.is_public: continue
            public_sessions.append(session)
        return public_sessions

    def update_session(self, session):
        token = session.token
        if self.read_session(token) is None: return
        session_dict = serialize_session(session)

        tests = {}
        if session.status != COMPLETED and session.status != ABORTED:
            tests[u"pending_tests"] = session.pending_tests
            tests[u"running_tests"] = session.running_tests
            tests[u"completed_tests"] = session.completed_tests

        tests[u"malfunctioning_tests"] = session.malfunctioning_tests
        self._tests_database.update_tests(token, tests)

        del session_dict[u"completed_tests"]
        del session_dict[u"running_tests"]
        del session_dict[u"pending_tests"]
        del session_dict[u"malfunctioning_tests"]

        for session in self._sessions:
            if session[u"token"] == token:
                self._sessions.remove(session)
                break
        self._sessions.append(session_dict)

    def delete_session(self, token):
        for session in self._sessions:
            if session[u"token"] == token:
                self._sessions.remove(session)
                break

        self._tests_database.delete_tests(token)
        self._results_database.delete_results(token)


    def find_tokens(self, fragment):
        tokens = []
        pattern = re.compile(u"^" + fragment)
        for session in self._sessions:
            if pattern.match(session[u"token"]) is not None:
                tokens.append(session[u"token"])
        return tokens